Implement parser for text-spacing and its longhands for auto and none

vitorroriz opened the pull request:
#### 45e26321c1e46e84ea2610ec9568783d365ffbe3
<pre>
Implement parser for text-spacing and its longhands for auto and none
<a href="https://bugs.webkit.org/show_bug.cgi?id=252068">https://bugs.webkit.org/show_bug.cgi?id=252068</a>
rdar://105094870
Reviewed by NOBODY (OOPS!).

Reference: <a href="https://github.com/w3c/csswg-drafts/issues/4246#issuecomment-1404738513">https://github.com/w3c/csswg-drafts/issues/4246#issuecomment-1404738513</a>

Implement a parser with longhands support for:
text-spacing: auto | none | &lt;text-autospace&gt; || &lt;text-spacing-trim&gt;

text-autospace: auto | no-autospace
text-spacing-trim: auto | space-all

We can then iterate from here adding support for &apos;normal&apos; in the
shorthand and for the remaining values in the longhands.

For the text-spacing shorthand, apart from appearing as a single token,
&apos;auto&apos; can also appear as belonging to one of the longhands in any order.
In case we parse an &apos;auto&apos; value and the token range is not exhauted,
there is a certain ambiguity if it belongs to the first or second
longhand, since they can come in any order.
Becuse all the other values of the longhands are unique, we can peek
into the next token to solve this ambiguity.

For the common case, where &apos;auto&apos; doesn&apos;t appear as the first token,
we try to re-use the parsing functions for the longhands. For that, we
need to divide the range into 2 subranges, one for each longhand.
We use again our knowledge of unique values for a longhand to identify
such subranges.

* LayoutTests/TestExpectations:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::textSpacingFromStyle):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
(WebCore::ShorthandSerializer::serializeTextSpacing const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeTextSpacing):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
(WebCore::CSSPropertyParserHelpers::isValueIDUniqueToTextSpacingTrimLonghand):
(WebCore::CSSPropertyParserHelpers::isValueIDUniqueToTextAutospaceLonghand):
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyle.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45e26321c1e46e84ea2610ec9568783d365ffbe3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107992 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17054 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40876 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117127 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116465 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111881 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18433 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8369 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100187 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113756 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13910 "Too many flaky failures: editing/deleting/smart-delete-002.html, editing/deleting/smart-delete-004.html, editing/deleting/smart-delete-paragraph-002.html, editing/deleting/smart-editing-disabled-win.html, editing/pasteboard/smart-paste-002.html, editing/pasteboard/smart-paste-004.html, editing/pasteboard/smart-paste-006.html, editing/pasteboard/smart-paste-008.html, editing/pasteboard/smart-paste-paragraph-003.html, editing/selection/doubleclick-whitespace-crash.html, editing/selection/ios/selection-handle-clamping-in-iframe.html, imported/w3c/web-platform-tests/html/semantics/popovers/popover-animated-hide-cleanup.tentative.html, platform/ipad/fast/viewport/empty-meta.html, platform/ipad/fast/viewport/meta-viewport-ignored.html, platform/ipad/fast/viewport/width-is-device-width.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97128 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41826 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95821 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28766 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83493 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9965 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30114 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10676 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7012 "Found 1 new test failure: fast/images/avif-image-decoding.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16116 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49704 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12252 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -1146,6 +1146,59 @@
+                "settings-flag": "cssTextSpacingEnabled",
+                "converter": "TextAutospace",
+                "parser-function": "consumeTextAutospace",
+                "parser-grammar-unused": "",
```
Please do include the grammar here.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -926,6 +929,22 @@ String ShorthandSerializer::serializeFont() const
+    result.append(serializeLonghandValue(longhandIndex(0, CSSPropertyTextAutospace)));
+    result.append(" ");
+    result.append(serializeLonghandValue(longhandIndex(1, CSSPropertyTextSpacingTrim)));
+    return result.toString();
```
Can you use `makeString()` here? It's more efficient than using `StringBuilder`

nt1m replied:
```
return makeString(
    serializeLonghandValue(longhandIndex(0, CSSPropertyTextAutospace),
    ' ',
    serializeLonghandValue(longhandIndex(1, CSSPropertyTextSpacingTrim)
);
```

pxlcoder commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -771,6 +771,105 @@ bool CSSPropertyParser::consumeFontVariantShorthand(bool important)
+    // From here, we know that first value is is not auto | none . Let's split the range in two subranges, one for each longhand consumer.
+    auto firstID = m_range.peek().id();
+    auto it = m_range;
+    bool isFirtValueFromTextSpacingTrimLongHand { false };
```
Typo: `isFirstValueFromTextSpacingTrimLongHand`.

pxlcoder commented on Source/WebCore/platform/text/TextSpacing.cpp:
```diff
@@ -0,0 +1,32 @@
+#include "TextSpacing.h"
+
+namespace WebCore {
+
```
Do you have plans to add to this file soon? Otherwise I think we can just remove it.

pxlcoder commented on Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
```diff
@@ -69,6 +69,8 @@ struct GreaterThanOrSameSizeAsStyleRareInheritedData : public RefCounted<Greater
     StyleColorScheme colorScheme;
 #endif
+    TextSpacingTrim textSpacingTrim;
+    TextAutospace textAutospace;
```
cc @heycam, who has been looking at moving things into / out of `StyleRareInheritedData`

nt1m replied:
I think this is appropriate in RareInheritedData, I don't expect super wide usage of these properties.

vitorroriz commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
```diff
@@ -436,6 +437,19 @@ RefPtr<CSSValueList> consumeCommaSeparatedListWithoutSingleValueOptimization(CSS
+{
+    // FIXME: add remaining values when specification is stable and we are parsing them.
+    return id == CSSValueSpaceAll;
+
```
remove extra empty line

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -917,6 +920,22 @@ String ShorthandSerializer::serializeFont() const
+    if (autospaceValue.valueID() == CSSValueAuto && trimValue.valueID() == CSSValueAuto)
+        return nameString(CSSValueAuto);
+    if (autospaceValue.valueID() == CSSValueNoAutospace && trimValue.valueID() == CSSValueSpaceAll)
+        return nameString(CSSValueNone);
```
```suggestion
    auto autospaceValueID = longhandValueID(longhandIndex(0, CSSPropertyTextAutospace));
    auto trimValueID = longhandValueID(longhandIndex(1, CSSPropertyTextSpacingTrim));
    if (autospaceValueID == CSSValueAuto && trimValueID == CSSValueAuto)
        return nameString(CSSValueAuto);
    if (autospaceValueID == CSSValueNoAutospace && trimValueID == CSSValueSpaceAll)
        return nameString(CSSValueNone);
```

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -763,6 +763,105 @@ bool CSSPropertyParser::consumeFontVariantShorthand(bool important)
+            return true;
+        }
+    }
+    // part II (longhands): <text-space-autospace> || <text-spacing-trim> (with fist token = auto or none)
+    if (firstValue && !m_range.atEnd()) {
```
This seems wrong, the grammar:
`normal | none | auto | <'text-autospace'> || <'text-spacing-trim'>`
should read as:

`normal | none | auto | [ <'text-autospace'> || <'text-spacing-trim'> ]`

according to https://drafts.csswg.org/css-values-4/#component-combinators

vitorroriz replied:
we are parsing `normal | none | auto | [ <'text-autospace'> || <'text-spacing-trim'> ]` indeed, being discussed on pm

vitorroriz commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -763,6 +763,105 @@ bool CSSPropertyParser::consumeFontVariantShorthand(bool important)
+    // Part I : auto | none
+    auto firstValue = consumeIdent<CSSValueNone, CSSValueAuto>(m_range);
+    if (firstValue && m_range.atEnd()) {
+        if (m_range.atEnd()) {
```
ops, this atEnd is redundant

vitorroriz committed:
Implement parser for text-spacing and its longhands for auto and none
https://bugs.webkit.org/show_bug.cgi?id=252068
rdar://105094870
Reviewed by NOBODY (OOPS!).

Reference: https://github.com/w3c/csswg-drafts/issues/4246#issuecomment-1404738513

Implement a parser with longhands support for:
text-spacing: auto | none | <text-autospace> || <text-spacing-trim>

text-autospace: auto | no-autospace
text-spacing-trim: auto | space-all

We can then iterate from here adding support for 'normal' in the
shorthand and for the remaining values in the longhands.

For the text-spacing shorthand, apart from appearing as a single token,
'auto' can also appear as belonging to one of the longhands in any order.
In case we parse an 'auto' value and the token range is not exhauted,
there is a certain ambiguity if it belongs to the first or second
longhand, since they can come in any order.
Becuse all the other values of the longhands are unique, we can peek
into the next token to solve this ambiguity.

For the common case, where 'auto' doesn't appear as the first token,
we try to re-use the parsing functions for the longhands. For that, we
need to divide the range into 2 subranges, one for each longhand.
We use again our knowledge of unique values for a longhand to identify
such subranges.

* LayoutTests/TestExpectations:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::textSpacingFromStyle):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
(WebCore::ShorthandSerializer::serializeTextSpacing const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeTextSpacing):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
(WebCore::CSSPropertyParserHelpers::isValueIDUniqueToTextSpacingTrimLonghand):
(WebCore::CSSPropertyParserHelpers::isValueIDUniqueToTextAutospaceLonghand):
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyle.h:
